### PR TITLE
Adds Random Seeds and a New Trait and Further Gene Datumizing

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Adds in random seeds.

Random seeds are...well, seeds that have highly randomized stats---additionally, the traits they possess (if any!) are also randomized---as is the type of plant they are. Probably their biggest novelty, however, is their ability to produce just about any reagent (minus `can_synth = 0` reagents) under the sun; you might find a meth plant one day, or a teslium plant the next.

You can find 2 packs of random seeds as contraband and can acquire two more every time you order an exotic seed crate.

Strange plant pics:

img1
img2

Also adds the very rare chance for a `hardmutate` to add in a random trait to a plant--currently only has a 3% of happening if you get the `hardmuate` proc (which isn't every mutate)

There's also a new mutation-only acquirable trait (or through random seeds) called "gaseous decomposition"; simply put, if a fruit is squashed (yes, you'll have to have the liquid contents gene for this to do anything), then the fruit will release a smoke cloud, spewing reagents everywhere; the size of the cloud scales with potency, ranging from very tiny to a moderate size (size at 100 potency isn't enough to cover a room the size of botany).

Another thing done is the conversion of weed growth rates and weed growth chance---both--over to core genes; this gives players the ability to modify/reduce/increase how many weeds form and how often. As a downside, however, mutating plants also increases/decreases weed rate/chance, so it's something that'll have to be managed now.

Lastly, fixes admin logging on the smoke reaction for smoke powder and the other reaction--arguments were swapped.

:cl: Fox McCloud
add: Adds in random botany seeds; never the same twice.
add: Adds in new trait that makes a grown release smoke when squashed
add: Weed rates and chances are now core seed genes
/:cl:

### CODE REVIEW

**Comment by XDTM** [comment]
imagine getting the plasmaman mutation toxin in your apples

**Comment by GunHog** [comment]
This can only end in wondrous, lovely **DISASTER!!!**

File: code/modules/hydroponics/grown/random.dm [hunk 1]
```diff
@@ -0,0 +1,29 @@
+//Random seeds; stats, traits, and plant type are randomized for each seed.
+
+/obj/item/seeds/random
+	name = "pack of strange seeds"
+	desc = "Mysterious seeds as strange as their name implies. Spooky"
+	icon_state = "seed-x"
```
Comment (TehZombehz) [inline code/modules/hydroponics/grown/random.dm]: Your customers won't know how they liiiived without brand X.

**Comment by Cheridan** [comment]
hooray someone finally used my alien plant sprites

**Comment by MisterTikva** [comment]
That gaseosut thingy is not needed as ayou already can make smoke grenades with separated chemicals

**Comment by FantasticFwoosh** [comment]
@MisterTikva however keeping the gas gene in would let people make smoke easier from the outlying components rather than diluting it with smoke chems.

**Comment by ma44** [comment]
I LIKE THIS

**Comment by FantasticFwoosh** [comment]
I hope this is either guranteed or atleast a variable for plant people loot between gatfruit and (placeholder cherry bombs until it was commented out).

>Either that or get a way to help get rare plant loot

**Comment by calzilla1** [comment]
Is it randomized every harvest or just initially?

**Comment by XDTM** [comment]
It **adds** traits/chems every harvest, if i read it correctly, which probably leads to a metric ton of random potentially op chems.

**Comment by ShadowDeath6** [comment]
You don't live until you fill a lemon with every traitor chem available

**Comment by MisterTikva** [comment]
So it doesn't mutate, traits are assigned on spawn and guess what i've got in my 3rd pack? 13% lexorin and stimulants.

**Comment by XDTM** [comment]
an omega weed with expanded storage can be filled with 840 u of juicy chems, per plant

**Comment by MisterTikva** [comment]
screenshot_72
Fifth pack.....

**Comment by MisterTikva** [comment]
Also it can add the same trait twice
screenshot_73

**Comment by MisterTikva** [comment]
I think it should always have at least one chemical in it, because it would be quite disappointing to not get a single chem gene out of all your seeds.

**Comment by Fox-McCloud** [comment]
The fact it can pick any reagent in the game (minus a few blacklisted ones), IMO, justifies it having a chance to literally give nothing. These aren't just for funsie time reagents; they're also to get additional traits/status that you might otherwise have a hard time getting.

Also, the traits/reagents do *not* randomize each harvest or for each grown.

**Comment by MisterTikva** [comment]
There is no traits that are hard to get if you have enough mutagen, they're all gotten with 1-2 mutations. As for the chems, maybe not guaranteed but increase the chance to something like 75-85% so you have to get unlucky to get no reagents, not almost 50/50.

**Comment by Fox-McCloud** [comment]
Alrighty, added some sanity so you can't have duplicate plant traits/reagent traits/plant type traits.

**Comment by Fox-McCloud** [comment]
converted `weed_rate` and `weed_chance` to core genes.

I think this pretty much converts all the current vars on seeds over to traits---though, hypothetically, we could start drilling down and convert things like nutrient absorption rate, light level tolerance, etc all over to specific attributes, but this is probably not important, for now.

**Comment by ma44** [comment]
Considering there's over 200 reagents, it would be rare and would require a lot of labor to get that one trait stored on a data disk.

I'll go try it out and see how it goes

**Comment by XDTM** [comment]
Remember that turning vars into traits means that every plant gets the best possible value for that var

**Comment by FantasticFwoosh** [comment]
@ma44 did you use robust harvest on the plant? Unless @XDTM's statement was a educated guess your both conflicting with your findings.

Try running 1 random plant on robust, another on ez grow, one on left4zed & another on natural compost and see if it has any effect to changing chems on harvest as per mutating the stats.

**Comment by Fox-McCloud** [comment]
Again, the seed's traits, genes, stats, etc are randomized when the seed packet is first created; anything produced from that packet will have the same stats/traits/genes unless you mutate it---even then, it's not going to randomize all the stats all over again.

**Comment by ma44** [comment]
damm :(

**Comment by XDTM** [comment]
Huh, my bad. From the code i'd thought that when you harvest you apply the New() to every new seed.

**Comment by Fox-McCloud** [comment]
@XDTM the seed is generated with no traits or genes when that happens--all its stats are copied over from the original.

**Comment by jughu** [comment]
With enough rng you can make blackpowder apples

**Comment by ExcessiveUseOfCobblestone** [comment]
Please go slow with the trend of putting everything behind gene manipulation. There's no fun when I can perfect literally every plant in every form possible.

As a botany player, not a big fan of being able to get Chems that are otherwise extremely difficult to obtain. Instead, why not add more basic Chems to plants so botany can be ghetto chem instead of being Chem+?

This is too big of a change to botany to put all in one go IMO.

**Comment by AdamElTablawy** [comment]
We do not need infinite initropidil smoke apples.

Botany is OP enough, and it's even worse on paradise. I really don't need flashbacks of the abomination that is paradise botany killing everyone every round, so taking things from your server and stapling them onto /tg/ is a bad idea when it doesn't even work for your own server.

**Comment by jughu** [comment]
Chems can be blacklisted

**Comment by jughu** [comment]
So far this auctally a cool idea and well would open your way to a lot of custom food as a comment said earlier it be cooler if they just include the base chems so you can ghetto chem your way into a cyanide bana or something like that

**Comment by ExcessiveUseOfCobblestone** [comment]
chems can also be made, and if you're going to blacklist all of the traitor chems you may as well do what I said earlier instead of making them better at chemistry than chemistry.

**Comment by jughu** [comment]
I mean amantin watermelons are still the strongest most unseen most unsuspicous and can be a insta death if you ate to much

**Comment by ExcessiveUseOfCobblestone** [comment]
that doesn't give license to slippery slope

**Comment by jughu** [comment]
I mean sure it can be better then chemistry but you have to take things in fact

It requires deparrmental interaction with cargo / or you need to hack the vendor
you must get lucky enough to get a usefull chem with hundreds useless ones outside there which can be tea grape juice beer and such
It takes effort getting a good final product out of it then you can easily produce the chems
I mean oddysues can pretty much replace a chemist job

**Comment by Iamgoofball** [comment]
departmental interaction is a complete hoax stop believinng it

**Comment by Fox-McCloud** [comment]
Could I get a review of this, please? 

As for the reagents that they can contain, there's a 40% chance a packet of seeds will contain no reagents at all---then it rolls again to determine how many reagents there will be, ranging from 0 to 2; so more than half the time, you'll end up getting no reagents.

Long term, I have no doubt that some more reagents will have to be blacklisted; thankfully the `can_synth` var works well for this; anything that is likely too powerful to be in plants/syringe guns is too powerful for syringe guns/plants.

All in all, they're a unique little thing for botany to fiddle with, but they're not, consistently, going to offer them a significant advantage over other departments; most of the time they're, frankly, going to get silly things.

File: code/modules/reagents/chemistry/holder.dm [hunk 2]
```diff
@@ -696,3 +696,12 @@ var/const/INJECT = 5 //injection
 		qdel(reagents)
 	reagents = new/datum/reagents(max_vol)
 	reagents.my_atom = src
+
+/proc/get_random_reagent_id()	// Returns a random reagent ID minus blacklisted reagents
+	var/list/random_reagents = list()
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/holder.dm]: reagents aren't created midround, so this list should be static (yes, still within the proc!) to save doing the looping every time.

(just make sure to put the loop under a ```if(!random_reagents.len)``` check)

**Comment by Fox-McCloud** [comment]
Like this, @RemieRichards 

And thanks, wasn't aware that static lists would function like that, inside a proc.

```
/proc/get_random_reagent_id()	
	var/static/list/random_reagents = list()
	if(!random_reagents.len)
		for(var/thing  in subtypesof(/datum/reagent))
			var/datum/reagent/R = thing
			if(initial(R.can_synth))
				random_reagents += initial(R.id)
	var/picked_reagent = pick(random_reagents)
	return picked_reagent
```

**Comment by RemieRichards** [comment]
@Fox-McCloud looks like it's tabbed too far (the entire thing, could just be github) but other than that, yes.
and yeah they do, it's because (and I found out from MSO, who found out from lummox) that static variables compile as global variables, regardless of where they are, global scope, in a definition, or in a proc.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

Yeah, I wouldn't be surprised if the tabbing on Git is screwed up; the actual code is fine.  And thanks, again, for that info; I'll keep that in mind for the future.

**Comment by XDTM** [comment]
Speaking of can_synth, i'd rather not sync the blacklist between odysseus and plants.

Plants can spawn roundstart with a random chem, which at that point is basically infinite and easily distributable.
Odysseuses need to have a sample of the chem they replicate to be able to get it, transferring the chems to beakers is a long chore, especially if you're doing it alone, and the chem production rate is fixed.

If i somehow manage to get nanomachines/tubercolosis/mutation toxins _and_ have an odysseus, i want to be able replicate them.

**Pull request merged**